src/Illuminate/View/Concerns/CachesComponents.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Illuminate\View\Concerns;
+
+use Illuminate\Container\Container;
+use Illuminate\Contracts\Support\Htmlable;
+use Illuminate\Contracts\View\CanBeCached;
+use Illuminate\Contracts\View\View;
+use Illuminate\Support\Facades\Cache;
+
+trait CachesComponents
+{
+    /**
+     * The cache controlling attributes for this component.
+     *
+     * @return array
+     */
+    protected function cacheAttributes()
+    {
+        return [
+            'x-cache',
+            'x-cache-key',
+            'x-cache-ttl',
+        ];
+    }
+
+    /**
+     * Determines if the component can be cached.
+     *
+     * @return bool
+     */
+    protected function canBeCached()
+    {
+        return $this instanceof CanBeCached;
+    }
+
+    /**
+     * Determines if the component should be cached.
+     *
+     * @return bool
+     */
+    protected function shouldBeCached()
+    {
+        return $this->attributes->get('x-cache', false) && $this->canBeCached();
+    }
+
+    /**
+     * The key used to cache the component.
+     *
+     * @param  array  $slots
+     * @return bool
+     */
+    protected function cacheKey(array $slots)
+    {
+        return $this->attributes->get('x-cache-key', static::class.sha1($this->attributes->except($this->cacheAttributes())->toHtml().serialize($slots)));
```

Comment:
As the `ComponentAttributeBag@get` uses the `value()` helper to resolve the default parameter, isn't a good idea to wrap the default key into a `Closure`? So it is lazy evaluated only when the developer does not provide a `x-cache-key` value.

I know code clarity is preferred over micro optimzations, and I didn't benchmark any differences, but calling `sha1()` and `serialize()` every time when the developer does provides a `x-cache-key` value could be avoided.